presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -460,6 +472,50 @@ private void checkTypesMatchForInsert(Insert insert, Scope queryScope, List<Colu
             }
         }
 
+        private void checkTypesMatchForNestedStructs(
+                Node node,
+                String errorMessage,
+                int i,
+                List<Type> columnRowTypes,
+                List<RowType.Field> expectedRowFields,
+                String fieldName)
+        {
+            if (columnRowTypes.size() != expectedRowFields.size()) {
+                throw new SemanticException(MISMATCHED_SET_COLUMN_TYPES,
+                        node,
+                        errorMessage + "Mismatch at column %d: '%s' has %d field(s) but expression has %d field(s)",
+                        i + 1,
+                        fieldName,
+                        expectedRowFields.size(),
+                        columnRowTypes.size());
+            }
+            for (int j = 0; j < columnRowTypes.size(); j++) {
```

Comment:
nit: Maybe we could rename `j` to `rowFieldIndex` if it makes sense.

Reply:
made the change